Code/GraphMol/MarvinParse/MarvinDefs.cpp

```diff
@@ -0,0 +1,2978 @@
+      << "</molecule>";
+
+  return out.str();
+}
+
+ptree MarvinMonomerSgroup::toPtree() const {
+  ptree out = MarvinMolBase::toPtree();
+
+  out.put("<xmlattr>.atomRefs", boost::algorithm::join(getAtomList(), " "));
+  out.put("<xmlattr>.title", title);
+  out.put("<xmlattr>.charge", charge);
+
+  return out;
+}
+
+MarvinSuperatomSgroupExpanded::MarvinSuperatomSgroupExpanded(
+    MarvinMolBase *parentInit) {
+  parent = parentInit;
+}
+
+MarvinSuperatomSgroupExpanded::~MarvinSuperatomSgroupExpanded() {}
+
+MarvinMolBase *MarvinSuperatomSgroupExpanded::copyMol(
+    std::string idAppendage) const {
+  auto outSgroup = new MarvinSuperatomSgroupExpanded(this->parent);
+  this->parent->sgroups.push_back(outSgroup);
+
+  outSgroup->molID = this->molID + idAppendage;
+  outSgroup->id = this->id + idAppendage;
+  outSgroup->title = this->title;
+
+  auto actualParent = this->parent;
+  while (actualParent->role() == "MultipleSgroup") {
+    actualParent = actualParent->parent;
+  }
+
+  // the only time this is to be called is when a mutliple group above it is
+  // being expanded and the new atoms and bonds have already been made
+  //  here we just need to find them and add refs to the new multiple sgroup
+
+  for (auto atomToCopy : atoms) {
+    auto atomIndex = actualParent->getAtomIndex(atomToCopy->id + idAppendage);
+    if (atomIndex < 0) {
+      throw FileParseException(
+          "unexpected error - new atom not found in parent");
+    }
+
+    outSgroup->atoms.push_back(actualParent->atoms[atomIndex]);
+  }
+
+  for (auto bondToCopy : bonds) {
+    auto bondIndex = actualParent->getBondIndex(bondToCopy->id + idAppendage);
+    if (bondIndex < 0) {
+      throw FileParseException(
+          "unexpected error - new bond not found in parent");
+    }
+
+    outSgroup->bonds.push_back(actualParent->bonds[bondIndex]);
+  }
+
+  return outSgroup;
+}
+
+std::string MarvinSuperatomSgroupExpanded::toString() const {
+  std::ostringstream out;
+
+  out << "<molecule molID=\"" << molID << "\" id=\"" << id
+      << "\" role=\"SuperatomSgroup\" atomRefs=\""
+      << boost::algorithm::join(getAtomList(), " ") << "\" title=\"" << title
+      << "\"/>";
+
+  return out.str();
+}
+
+ptree MarvinSuperatomSgroupExpanded::toPtree() const {
+  ptree out = MarvinMolBase::toPtree();
+
+  out.put("<xmlattr>.atomRefs", boost::algorithm::join(getAtomList(), " "));
+  out.put("<xmlattr>.title", title);
+  return out;
+}
+
+std::string MarvinSuperatomSgroupExpanded::role() const {
+  return std::string("SuperatomSgroupExpanded");
+}
+
+bool MarvinSuperatomSgroupExpanded::hasAtomBondBlocks() const { return false; }
+
+MarvinSuperatomSgroup::MarvinSuperatomSgroup(MarvinMolBase *parentInit) {
+  parent = parentInit;
+}
+
+MarvinSuperatomSgroup::~MarvinSuperatomSgroup() {
+  for (auto &attachmentPoint : attachmentPoints) {
+    delete attachmentPoint;
+  }
+
+  for (auto &atom : atoms) {
+    delete atom;
+    atom = nullptr;
```

Comment:
Whare are you setting this to nullptr, but not doing the same thing above for attachmentPoint?
Doesn't seem necessary.

It would be easier/better to just `clear()` all the vectors after you `delete` their elements

Reply:
This was to ensure that the atom could not be double deleted.  This code is no longer there - it has been replaced by unique_ptrs and there is no explicit delete call